OpenWeatherApp/Providers/CityProviderRestAPI.swift

```diff
@@ -0,0 +1,35 @@
+//
+//  CityProviderRestAPI.swift
+//  OpenWeatherApp
+//
+//  Created by Ivan De Martino on 2/12/19.
+//  Copyright © 2019 Ivan De Martino. All rights reserved.
+//
+
+import Foundation
+
+struct CityProviderRestAPI: CityProviderType {
+  
+  private let apiURL = "https://api.openweathermap.org/data/2.5/weather"
+  private let apiKey = "acedb8e3b8769854a49f965fa13752e2"
+  
+  func fetch(with searchTerm: String, completion: @escaping (City) -> ()) {
+    
+    guard let url = URL(string: "\(apiURL)?q=\(searchTerm)&APPID=\(apiKey)") else { return }
```

Comment:
Bom vamos por partes. 
```Swift
let url: URL = {
            if endpoint.path.isEmpty {
                return endpoint.baseURL
            } else {
                return endpoint.baseURL.appendingPathComponent(endpoint.path)
            }
        }()
```
Eu entendi o que o código faz, mas só estranhei a sintaxe, os () no final.

Reply:
Fiz uns testes no playgrouds e acho que consegui entender. Sem os () no final ele so retorna uma referencia pra uma funcao e com os () ele executa e retorna o valor, é isso?

Reply:
Esse parametro resolvingAgainstBaseURL eu nao entendi o que faz: 
```Swift
var urlComponents = URLComponents(url: url, resolvingAgainstBaseURL: false)
```
Vc passou false, para nao usar os componentes da url absoluta?

Reply:
> Bom vamos por partes.
> ...
> Eu entendi o que o código faz, mas só estranhei a sintaxe, os () no final.

Basicamente criamos uma closure que será executada quando usarmos a constante `url`. É como criar uma `lazy var` porém como constante.

> Fiz uns testes no playgrouds e acho que consegui entender. Sem os () no final ele so retorna uma referencia pra uma funcao e com os () ele executa e retorna o valor, é isso?

Exatamente! Essa é a ideia, sem o `()` ele é um objeto do tipo closure.

> var urlComponents = URLComponents(url: url, resolvingAgainstBaseURL: false)

Segundo a documentação:
```
resolve
Controls whether the URL should be resolved against its base URL before parsing. If true, and if the url parameter contains a relative URL, the original URL is resolved against its base URL before parsing by calling the absoluteURL method. Otherwise, the string portion is used by itself.
```

Reply:
Essa parte eu estranhei também
```Swift
200..<300 ~= httpResponse.statusCode
```
Eu sei que as respostas HTTP da família 200 são OK e da 300 são redirecionamento. Eu só não entendi essa sintaxe. Seria respostas entre 200 e 300? Esse código recebendo um valor eu não entendi
```Swift  
~= 
````

Reply:
O operador `~=` funciona como um `contains` em um range. Ou seja, tudo que estiver entre `200 e 299` será considerado como sucesso.

[Documentação](https://developer.apple.com/documentation/swift/range/2428743)

Reply:
No caso de algum Redirect status 300, nesse caso será considerado uma falha. Isso seria uma boa pratica se tratando de consumir uma API?

Reply:
Eu consideraria uma falha dado que o range de sucesso é o `200`.

**Vamos imaginar a seguinte situação:**
O backend é terceirizado, um endpoint foi modificado e ao invés de retornar `200` retorna `301 - Moved`. 
Se não considerarmos como falha um `30X`, nunca mudaremos no lado do client para endpoint correto.

Reply:
Entendi. Eu sei que vc passa um Enum (APIRequesterError) que implementa Error com o case invalid response (pra esse caso). Poderia ser uma struct que implementa error e recebe o status code no init, para criar um error mais descritivo?

Reply:
Sim, é possivel mas não comun.

Normalmente o que vejo muito é o uso de [associated values](https://docs.swift.org/swift-book/LanguageGuide/Enumerations.html#ID148):

```Swift
enum APIRequesterError: Error {
   case httpError(Int)
   case invalidResponse

   var description: String {
       switch self {
             case httpError(let statusCode):
                  return "We had a HTTP error with status code: \(statusCode)"
             case invalidResponse:
                  return "Invalid response" 
       }
   }
}
```